clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/AggKafkaConsumerService.java

```diff
@@ -177,80 +186,104 @@ protected static Runnable getStuckConsumerDetectionAndRepairRunnable(
+            logger.error("Hit exception while killing the stuck ingestion task for store version: {}", vt, e);
+          }
+        });
+        if (!hasRepairedSomeIngestionTask.get()) {
+          if (!REDUNDANT_LOGGING_FILTER.isRedundantException(STUCK_CONSUMER_MSG)) {
+            logger.warn(STUCK_CONSUMER_MSG);
           }
+          stuckConsumerRepairStats.recordRepairFailure();
         }
       }
 
-      AtomicBoolean hasRepairedSomeIngestionTask = new AtomicBoolean(false);
-      versionTopicIngestionTaskMappingForNonExistingTopic.forEach((vt, sit) -> {
-        try {
-          LOGGER.warn(
-              "The ingestion topics (version topic) are not healthy for "
-                  + "store version: {}, will kill the ingestion task to try to unblock shared consumer",
-              vt);
-          /**
-           * The following function call will interrupt all the stuck {@link org.apache.kafka.clients.producer.KafkaProducer#send} call
-           * to non-existing topics.
-           */
-          sit.closeVeniceWriters(false);
-          killIngestionTaskRunnable.accept(vt);
-          hasRepairedSomeIngestionTask.set(true);
-          stuckConsumerRepairStats.recordIngestionTaskRepair();
-        } catch (Exception e) {
-          LOGGER.error("Hit exception while killing the stuck ingestion task for store version: {}", vt, e);
-        }
-      });
-      if (!hasRepairedSomeIngestionTask.get()) {
-        if (!REDUNDANT_LOGGING_FILTER.isRedundantException(STUCK_CONSUMER_MSG)) {
-          LOGGER.warn(STUCK_CONSUMER_MSG);
+      StringBuilder stringBuilder = new StringBuilder();
+      long now = time.getMilliseconds();
+      // Detect and log any subscribed topic partitions that are not polling records
+      for (Map.Entry<String, AbstractKafkaConsumerService> consumerService: kafkaServerToConsumerServiceMap
+          .entrySet()) {
+        Map<PubSubTopicPartition, Long> staleTopicPartitions =
+            consumerService.getValue().getStaleTopicPartitions(consumerPollTrackerStaleThresholdMs);
+        if (!staleTopicPartitions.isEmpty()) {
+          stringBuilder.append(CONSUMER_POLL_WARNING_MESSAGE_PREFIX);
+          stringBuilder.append(consumerService.getKey());
+          for (Map.Entry<PubSubTopicPartition, Long> staleTopicPartition: staleTopicPartitions.entrySet()) {
+            stringBuilder.append("\n topic: ");
+            stringBuilder.append(staleTopicPartition.getKey().getTopicName());
+            stringBuilder.append(" partition: ");
+            stringBuilder.append(staleTopicPartition.getKey().getPartitionNumber());
+            stringBuilder.append(" stale for: ");
+            stringBuilder.append(now - staleTopicPartition.getValue());
+            stringBuilder.append("ms");
+          }
+          logger.warn(stringBuilder.toString());
+          // clear the StringBuilder to be reused for next reporting cycle
+          stringBuilder.setLength(0);
```

Comment:
This function is already quite lengthy, do you think we could abstract the new functionality into a new function?